packages/lib-classifier/src/store/FeedbackStore.js

```diff
@@ -1,45 +1,94 @@
 import { autorun } from 'mobx'
-import { addDisposer, getRoot, onAction, types } from 'mobx-state-tree'
+import { addDisposer, addMiddleware, getRoot, onAction, types } from 'mobx-state-tree'
+import { flatten } from 'lodash'
 
 import helpers from './feedback/helpers'
 import strategies from './feedback/strategies'
 
 const FeedbackStore = types
   .model('FeedbackStore', {
     isActive: types.optional(types.boolean, false),
-    rules: types.map(types.frozen({}))
+    rules: types.map(types.frozen({})),
+    showModal: types.optional(types.boolean, false)
   })
+  .volatile(self => ({
+    onHide: () => true
+  }))
+  .views(self => ({
+    get hideSubjectViewer () {
+      return flatten(Array.from(self.rules.values()))
+        .some(rule => rule.hideSubjectViewer)
+    },
+    get messages () {
+      return flatten(Array.from(self.rules.values()))
+        .map(rule => {
+          if (rule.success && rule.successEnabled) {
+            return rule.successMessage
+          } else if (!rule.success && rule.failureEnabled) {
+            return rule.failureMessage
+          }
+        }).filter(Boolean)
+    }
+  }))
   .actions(self => {
-    function afterAttach () {
-      createSubjectObserver()
-      createClassificationObserver()
+    function setOnHide (onHide) {
+      self.onHide = onHide
     }
 
-    function createSubjectObserver () {
-      const subjectDisposer = autorun(() => {
-        const subject = getRoot(self).subjects.active
-        if (subject) {
-          self.reset()
-          self.createRules(subject)
-        }
-      })
-      addDisposer(self, subjectDisposer)
+    function afterAttach () {
+      createClassificationObserver()
+      createSubjectMiddleware()
+      createSubjectObserver()
     }
 
     function createClassificationObserver () {
       const classificationDisposer = autorun(() => {
         onAction(getRoot(self).classifications, (call) => {
           if (call.name === 'completeClassification') {
             const annotations = getRoot(self).classifications.currentAnnotations
-            for (const value of annotations.values()) {
-              self.update(value)
-            }
+            annotations.forEach(annotation => self.update(annotation))
           }
         })
       })
       addDisposer(self, classificationDisposer)
     }
 
+    function onSubjectAdvance (call, next, abort) {
+      const shouldShowFeedback = self.isActive && self.messages.length && !self.showModal
+      if (shouldShowFeedback) {
+        abort()
+        const onHide = getRoot(self).subjects.advance
```

Comment:
One of my dislikes about this implementation is that it's recursive. Calling `subjects.advance` at the end of feedback will trigger this handler to run again. We've hacked around that by checking to see if there's a modal open.

Reply:
The reason I don't like it is that it's possible to accidentally put the code into a never-ending loop where closing the feedback modal opens it again and you never get to the next subject.